### ISSUE TITLE
torch.ldexp backward returns zero gradient w.r.t. mantissa for negative integer exponents

### ISSUE THREAD

**Issue by OE-GOD** [pytorch/pytorch#187460]
### 🐛 Describe the bug

`torch.ldexp(input, other)` computes `input * 2**other`, so the gradient w.r.t. `input` is `2**other`. When `other` is an integer tensor with a negative value, the backward pass returns **0** instead, while the forward pass is correct:

```python
import torch
m = torch.tensor([0.5], requires_grad=True)
e = torch.tensor([-1], dtype=torch.int32)
torch.ldexp(m, e).backward()
print(m.grad)   # tensor([0.])  — should be 0.5

# float exponent is correct:
m = torch.tensor([0.5], requires_grad=True)
torch.ldexp(m, torch.tensor([-1.0])).backward()
print(m.grad)   # tensor([0.5000])
```

The wrong gradient exactly equals `2**e` evaluated in **integer** arithmetic (floors to 0 for any negative `e`; correct for `e >= 0`), so the backward multiplier is being computed with an integer power instead of the real-valued `2.0**e` the forward uses:

| e (int32) | forward `ldexp(1, e)` | grad w.r.t. mantissa | correct `2**e` |
|---|---|---|---|
| -1 | 0.5 | **0.0** | 0.5 |
| -2 | 0.25 | **0.0** | 0.25 |
| -3 | 0.125 | **0.0** | 0.125 |
| 0 | 1.0 | 1.0 | 1.0 |
| 2 | 4.0 | 4.0 | 4.0 |

Finite differences confirm the true derivative: `(ldexp(0.5+h, -1) - ldexp(0.5-h, -1)) / 2h = 0.5`. Forward and backward therefore disagree.

**Impact:** silent — differentiating through `ldexp(x, e)` with a negative integer exponent yields a zero gradient, so the parameter simply stops training, with no error.

**Suggested fix:** cast `other` to the input's floating dtype before computing `2**other` in the backward formula (matching the forward).

### Versions

Reproduced on torch 2.12.0, CPU (arm64, macOS).

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

**Comment by OE-GOD** [comment]
Closing as a duplicate of #186556, which reported this same `torch.ldexp` negative-exponent gradient bug on 2026-06-07. Apologies for the duplicate — my search used narrower terms and missed it. #186556 is the canonical report.
